Generate streams

EPashkin opened the pull request:
cc @GuillaumeGomez, @sdroege

GuillaumeGomez commented:
👍

sdroege commented on Gir.toml:
```diff
@@ -72,6 +94,14 @@ status = "generate"
+    [[object.function]]
+    name = "read_line"
+    #return vec of char*
+    ignore = true
```
That should just return a `Vec<u8>` and I thought gir handles that by now (note that the element type is set to guint8 in the gir file, it's not utf8).

EPashkin replied:
Gir generated good looking code:
```rust
    fn read_line<'a, P: Into<Option<&'a Cancellable>>>(&self, cancellable: P) -> Result<(Vec<u8>, usize), Error> {
        let cancellable = cancellable.into();
        let cancellable = cancellable.to_glib_none();
        unsafe {
            let mut length = mem::uninitialized();
            let mut error = ptr::null_mut();
            let ret = ffi::g_data_input_stream_read_line(self.to_glib_none().0, &mut length, cancellable.0, &mut error);
            if error.is_null() { Ok((FromGlibPtrContainer::from_glib_full(ret), length)) } else { Err(from_glib_full(error)) }
        }
    }
```
but compiling get 2 errors:
```
error[E0277]: the trait bound `u8: glib::translate::Ptr` is not satisfied
   --> src\auto\data_input_stream.rs:145:38
    |
145 |             if error.is_null() { Ok((FromGlibPtrContainer::from_glib_full(ret), length)) } else { Err(from_glib_full(error)) }
    |                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `glib::translate::Ptr` is not implemented for `u8`
    |
    = note: required because of the requirements on the impl of `glib::translate::FromGlibPtrContainer<u8, *mut u8>` for `std::vec::Vec<u8>`
    = note: required by `glib::translate::FromGlibPtrContainer::from_glib_full`

error[E0277]: the trait bound `u8: glib::translate::FromGlibPtrArrayContainerAsVec<u8, *mut u8>` is not satisfied
   --> src\auto\data_input_stream.rs:145:38
    |
145 |             if error.is_null() { Ok((FromGlibPtrContainer::from_glib_full(ret), length)) } else { Err(from_glib_full(error)) }
    |                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `glib::translate::FromGlibPtrArrayContainerAsVec<u8,
 *mut u8>` is not implemented for `u8`
    |
    = note: required because of the requirements on the impl of `glib::translate::FromGlibPtrContainer<u8, *mut u8>` for `std::vec::Vec<u8>`
    = note: required by `glib::translate::FromGlibPtrContainer::from_glib_full`
```

sdroege replied:
Hmm, possibly a bug in glib/translate.rs. Exactly that impl should be used here and should be usable, unclear why it isn't. The `glib::translate::Ptr` requirement looks dubious.

Can you create a ticket?

EPashkin replied:
Added https://github.com/gtk-rs/glib/issues/275

sdroege commented on Gir.toml:
```diff
@@ -95,6 +125,14 @@ status = "generate"
+    [[object.function]]
+    name = "splice_async"
+    #g_io_stream_splice_finish accept only 2 parameters instead normal 3 (no source_object)
+    ignore = true
```
@antoyo ↑

antoyo replied:
Do I need to update `gir` or something?

sdroege commented on src/auto/buffered_input_stream.rs:
```diff
@@ -0,0 +1,153 @@
+
+    fn get_buffer_size(&self) -> usize;
+
+    fn peek(&self, buffer: &[u8], offset: usize) -> usize;
```
The buffer needs to be mutable, the function reads into this buffer

EPashkin replied:
Seems this function need be manual :(

sdroege commented on src/auto/converter.rs:
```diff
@@ -0,0 +1,47 @@
+}
+
+impl<O: IsA<Converter>> ConverterExt for O {
+    fn convert(&self, inbuf: &[u8], outbuf: &[u8], flags: ConverterFlags) -> Result<(ConverterResult, usize, usize), Error> {
```
outbuf must be mutable

EPashkin replied:
Will made all Converter manual

sdroege commented on src/auto/functions.rs:
```diff
@@ -300,20 +312,42 @@ pub fn networking_init() {
-//    unsafe { TODO: call ffi::g_pollable_stream_read() }
-//}
+#[cfg(any(feature = "v2_34", feature = "dox"))]
+pub fn pollable_stream_read<'a, P: IsA<InputStream>, Q: Into<Option<&'a Cancellable>>>(stream: &P, buffer: &[u8], blocking: bool, cancellable: Q) -> Result<isize, Error> {
```
This should not be in functions, but in InputStream (and the other one in OutputStream). Also calling this with blocking=false on a non-GPollableInputStream is undefined behaviour, thus unsafe. I would just not expose this at all, it's C convenience stuff with footguns

sdroege replied:
Also the buffer needs to be writable

EPashkin replied:
Ok I ignore this function.

EPashkin replied:
Maybe ignore `pollable_stream_write` too? PollableStreams not generated.

sdroege replied:
Yes, I meant both

sdroege commented on src/auto/input_stream.rs:
```diff
@@ -0,0 +1,248 @@
+    #[cfg(any(feature = "v2_44", feature = "dox"))]
+    fn read_all_async<'a, P: Into<Option<&'a Cancellable>>, Q: Fn(Result<usize, Error>) + Send + Sync + 'static>(&self, buffer: &[u8], io_priority: i32, cancellable: P, callback: Q);
+
+    fn read_async<'a, P: Into<Option<&'a Cancellable>>, Q: Fn(Result<(), Error>) + Send + Sync + 'static>(&self, buffer: &[u8], io_priority: i32, cancellable: P, callback: Q);
```
Buffer in all the read functions must be mutable

sdroege commented on src/auto/input_stream.rs:
```diff
@@ -0,0 +1,248 @@
+
+    fn is_closed(&self) -> bool;
+
+    fn read<'a, P: Into<Option<&'a Cancellable>>>(&self, buffer: &[u8], cancellable: P) -> Result<isize, Error>;
```
Buffer in all the read functions must be mutable

sdroege commented on src/auto/input_stream.rs:
```diff
@@ -0,0 +1,248 @@
+    #[cfg(any(feature = "v2_44", feature = "dox"))]
+    fn read_all_async<'a, P: Into<Option<&'a Cancellable>>, Q: Fn(Result<usize, Error>) + Send + Sync + 'static>(&self, buffer: &[u8], io_priority: i32, cancellable: P, callback: Q);
+
+    fn read_async<'a, P: Into<Option<&'a Cancellable>>, Q: Fn(Result<(), Error>) + Send + Sync + 'static>(&self, buffer: &[u8], io_priority: i32, cancellable: P, callback: Q);
```
And this buffer must even stay alive until the async callback is called

EPashkin replied:
There way to do it?

sdroege replied:
With autogenerated code? No, not right now

EPashkin replied:
In manual code?

sdroege replied:
Yes

sdroege commented on src/auto/output_stream.rs:
```diff
@@ -0,0 +1,304 @@
+    fn is_closing(&self) -> bool;
+
+    //#[cfg(any(feature = "v2_40", feature = "dox"))]
+    //fn printf<'a, P: Into<Option<&'a Cancellable>>>(&self, cancellable: P, error: &mut Error, format: &str, : /*Unknown conversion*//*Unimplemented*/Fundamental: VarArgs) -> Option<usize>;
```
Wat! I didn't know this function exists. Good to know :)

sdroege commented on src/auto/output_stream.rs:
```diff
@@ -0,0 +1,304 @@
+    #[cfg(any(feature = "v2_44", feature = "dox"))]
+    fn write_all_async<'a, P: Into<Option<&'a Cancellable>>, Q: Fn(Result<usize, Error>) + Send + Sync + 'static>(&self, buffer: &[u8], io_priority: i32, cancellable: P, callback: Q);
+
+    fn write_async<'a, P: Into<Option<&'a Cancellable>>, Q: Fn(Result<(), Error>) + Send + Sync + 'static>(&self, buffer: &[u8], io_priority: i32, cancellable: P, callback: Q);
```
Buffer must stay alive until the async callback (or rather the finish method) is called

sdroege commented:
Mostly looks ok, just that all read functions need a mutable buffer, all read/write/etc async functions need their buffer to stay alive until the async callback is called, and various other minor things.

EPashkin commented on src/auto/input_stream.rs:
```diff
@@ -0,0 +1,170 @@
+        unsafe {
+            ffi::g_input_stream_read_bytes_async(self.to_glib_none().0, count, io_priority, cancellable.0, Some(callback), Box::into_raw(user_data) as *mut _);
+        }
+    }
```
Seems this function generated wrong, see https://developer.gnome.org/gio/stable/GInputStream.html#g-input-stream-read-bytes-async
Don't understand how user get `glib::Bytes` from it 
cc @antoyo

sdroege replied:
g_input_stream_read_bytes_finish() gives you the GBytes

EPashkin commented:
Made `read_bytes_async` manual, 
There still question, how to "buffer must even stay alive until the async callback is called" even in manual code and this related to almost all async functions :(

EPashkin commented:
Make some async functions manual, moved entire `file_input_stream.rs` to manual as it have only two function and one of its `query_info_async` - wrong.
Will add other changes as separate commits, to simplify checking.

antoyo commented:
@EPashkin So, if I understand correctly, what's missing is the return value (`Ok` value of the `Result`), right?

EPashkin commented:
Yes.
In `Result`-returning functions we normally skip it only if return type `void` or `bool`

EPashkin commented:
@sdroege All async IMHO "fixed"

EPashkin commented:
But forgot make some buffers mutable, still fixing :(

EPashkin commented:
Done too.
Only last problem is "buffer must even stay alive until the async callback is called" for all async operation.

sdroege commented:
> There still question, how to "buffer must even stay alive until the async callback is called" even in manual code and this related to almost all async functions :(

Look at the GStreamer Buffer function I linked elsewhere: https://github.com/sdroege/gstreamer-rs/blob/963557b79f5d3a91db944ed953a8c9fcab78a6ba/gstreamer/src/buffer.rs#L100

EPashkin commented:
Sorry, I don't understand.

sdroege commented:
You can take any `AsMut<[u8]>` here, but let's ignore that part for now and take `Vec<u8>`. You need to box it in the async function and then make that part of the `user_data` (https://github.com/sdroege/gstreamer-rs/blob/963557b79f5d3a91db944ed953a8c9fcab78a6ba/gstreamer/src/buffer.rs#L76-L81). In the finish function you would steal it from the `user_data` (i.e. move it out of there!), in the destroy_notify for the user_data you would drop it (like https://github.com/sdroege/gstreamer-rs/blob/963557b79f5d3a91db944ed953a8c9fcab78a6ba/gstreamer/src/buffer.rs#L67) if it was not stolen (see before).

EPashkin commented:
Is I not wrong on you just save pointer and size without blocking owner of memory do what his want. 
So I don't see how this protect this memory block from free and pollute with other data.

sdroege commented:
> Is I not wrong on you just save pointer and size without blocking owner of memory do what his want.
So I don't see how this protect this memory block from free and pollute with other data.

The function takes ownership of the data passed in, and only ever releases it from the destroy_notify or when the user releases it after calling the finish function.

EPashkin commented:
So you meant that [`write_async`](https://github.com/gtk-rs/gio/pull/69/files#diff-7c71e04c0e2366e61ffa11af9cc3e345R49) must take `buffer: Vec<u8>` instead `buffer: &[u8]`,
and always drop it in `write_async_trampoline` ?

sdroege commented:
Yes, look at the code I linked above. Either `Vec<u8>` or `AsRef<[u8]>`.

EPashkin commented:
`AsRef<[u8]>` IMHO don't transfer ownership inside function.

sdroege commented:
> AsRef<[u8]> IMHO don't transfer ownership inside function.

(I was omitting some trait bounds for simplicity, see the code I linked for what exactly you need)
`fn bla<T: AsRef<[u8]> + Send + 'static>(t: T)` does transfer ownership, `fn bla<T: AsRef<[u8]> + Send + 'static>(t: &T)` doesn't.

EPashkin commented:
Oh, I missed `b:&T` vs `b:T` :(
Still without `+'static`, rust allow this:
```rust
    fn tst(&self) {
        let mut buf = vec![1u8,2,3,4,5,7];
        self.write_async(&buf, 1, None, callback);
        buf[0]=2;
    }
    fn write_async<'a, T: AsRef<[u8]> + Send /*+ 'static*/, P: Into<Option<&'a Cancellable>>, Q: Fn(Result<isize, Error>) + Send + Sync + 'static>(&self, buffer: T, io_priority: i32, cancellable: P, callback: Q) {
```

With `+'static` it don't need drop, but IMHO still allowed.
Get `internal compiler error` on trying repeat example :(
Can you show working example of `AsRef<[u8]> + 'static` usage (actual version of your file have `pub fn from_vec(vec: Vec<u8>) -> Option<Self> {` ) ?

EPashkin commented:
I think that for async operations with buffer it need be returned in callback.
I get this compiled as example, but not tested if it works.
This ugly :(
```
    fn write_async<'a, P: Into<Option<&'a Cancellable>>, Q: Fn(Vec<u8>, Result<isize, Error>) + Send + Sync + 'static>(&self, buffer: Vec<u8>, io_priority: i32, cancellable: P, callback: Q) {
        let cancellable = cancellable.into();
        let cancellable = cancellable.to_glib_none();
        let count = buffer.len() as usize;
        let mut user_data: Box<(Box<Fn(Vec<u8>, Result<isize, Error>) + Send + Sync + 'static>, Vec<u8>)> = Box::new((Box::new(callback), buffer));
        let ptr = user_data.1.as_mut_ptr();
        extern "C" fn write_async_trampoline(_source_object: *mut gobject_ffi::GObject, res: *mut ffi::GAsyncResult, user_data: glib_ffi::gpointer)
        {
            callback_guard!();
            unsafe {
                let mut error = ptr::null_mut();
                let ret = ffi::g_output_stream_write_finish(_source_object as *mut _, res, &mut error);
                let result = if error.is_null() { Ok(ret) } else { Err(from_glib_full(error)) };
                let box_: Box<(Box<Fn(Vec<u8>, Result<isize, Error>) + Send + Sync + 'static>, Vec<u8>)> = Box::from_raw(transmute(user_data));
                let box_ = *box_;
                let (cb, buf) = box_;
                cb(buf, result);
            }
        }
        let callback = write_async_trampoline;
        unsafe {
            ffi::g_output_stream_write_async(self.to_glib_none().0, ptr, count, io_priority, cancellable.0, Some(callback), Box::into_raw(user_data) as *mut _);
        }
    }

```

sdroege commented:
As I said above, and also do in the linked code, you need `'static` (and `Send`). Otherwise exactly the problem you mention is possible, and that's not good :)

sdroege commented:
Why do you think it must be returned in callback? It can, but it does not have to. You can as well drop it from the trampoline (and in case of the read functions, you get it back from the finish() function and that you indeed want to pass to the callback).

sdroege commented:
Your code should work though, but you could also use `AsRef<[u8]>` here. Then for example the `Chunk` type from hyper could also be directly passed here instead of copying to a `Vec` first.

EPashkin commented:
`+'static` still allow pass reference, so `AsRef` is not safe anyway :(
```
static buf0:&[u8] = &[1,2,3];
fn callback(ret: Result<isize, Error>) {}
//in trait
fn tst(&self) {
        self.write_async(&buf0, 1, None, callback);
        let b = buf0[0];//=2;
    }
    fn write_async<'a, T: AsRef<[u8]> + Send + 'static, P: Into<Option<&'a Cancellable>>, Q: Fn(Result<isize, Error>) + Send + Sync + 'static>(&self, buffer: T, io_priority: i32, cancellable: P, callback: Q) {
```

Also it seems don't work good with `&mut [u8]` (but at minimum it prevent reference passing),
so readers can't be done this way.

Maybe better way add some type, that works like `Arc<RefCell<T>>` but allow block access while async action in progress.

EPashkin commented:
Forgot to say that [`write_bytes_async`](https://github.com/gtk-rs/gio/pull/69/files#diff-7c71e04c0e2366e61ffa11af9cc3e345R71) will need "lock" `glib::Bytes` and `+'static` IMHO don't make sense for this case.

sdroege commented:
I'll write you an async read and write function tomorrow, then we have something to discuss about :)

sdroege commented:
Also generally, you *must* not use `&[u8]` for the async read/write functions, unless they have a `'static` lifetime. Preparing something now :)

sdroege commented:
@EPashkin https://github.com/sdroege/gio-rs/commit/10dd01f30809cc72f768dc70bb3c8bed9b4e3f88 This is what I meant, see the commit message for the details of all changes (I also noticed a bug with all async methods currently :( see https://github.com/gtk-rs/gir/issues/508 ).

sdroege commented:
Tests for this can be easily written around the memory input/output streams btw

EPashkin commented:
Easy for async functions? 😉

EPashkin commented:
There was [build fix in `read_all_async`](https://github.com/gtk-rs/gio/commit/8444a1378a7106a361bacd532103cf13bfcfd8e3)

Some changes in `read_all_async` that I propose:
[don't modify buffer but return readed instead](https://github.com/gtk-rs/gio/commit/ce88681a63fd5504dd28cc047df5c7876a1e4a0f)
[Use `AsMut<[u8]>` instead array](https://github.com/gtk-rs/gio/commit/ae24af418803168f77c4ee06cea6c5c271faa4ed)

EPashkin commented:
@sdroege Updated `read_all_async` to use `Fn(Vec<u8>, Option<Error>)`

sdroege commented:
Seems ok now, will go over it another time in detail later/tomorrow

EPashkin commented:
Ok, so this question is for later.
We will use `AsMut<[u8]>` for this function?
What we pass to clouse then: `AsMut<[u8]>` or actual type?

sdroege commented:
For which function?

EPashkin commented:
Now only for `read_all_async`

sdroege commented:
The read ones must be Vec as you need set_len

EPashkin commented:
You right, it need for all read functions. If we don't want add special trait for this 😉

EPashkin commented:
Changed `InputStream::read_all`, maybe error tuple should reversed?
Added tests for `MemoryInputStream`.
Found that `new_from_data` and `add_data` can't be done in safe way,
as internally `gio` creates `glib::Bytes` from buffer with passing same pointer to `destroy` function,
so way used in `Bytes::from_owned` impossible.
@sdroege can you check this when you have free time?

sdroege commented:
Yes, only the functions that take a GBytes work with bindings (unless you copy into new malloc'd memory or similar but don't)

EPashkin commented:
Then maybe add version "v2_34" for `MemoryInputStream` to limit it usage?
And for `MemoryOutputStream` it will be "v2_36"

sdroege commented:
I guess you could in theory get such a stream from other API and only use the generic stream API on it

EPashkin commented:
Ok.

EPashkin commented:
@GuillaumeGomez Please restart Travis here.

GuillaumeGomez commented:
Done.

sdroege commented:
Might want to wait for https://github.com/gtk-rs/gir/pull/509 (it also affects the manual code, that would need some updating)

sdroege commented on src/lib.rs:
```diff
@@ -29,13 +38,26 @@ pub use glib::{
+use converter::Converter;
+use file_i_o_stream::FileIOStream;
+use file_input_stream::FileInputStream;
+use file_output_stream::FileOutputStream;
```
Why not *?

sdroege commented on src/input_stream.rs:
```diff
@@ -0,0 +1,180 @@
+    fn read_all<'a, P: Into<Option<&'a Cancellable>>>(&self, buffer: Vec<u8>, cancellable: P) -> Result<Vec<u8>, (Error, Vec<u8>)>;
+
+    #[cfg(any(feature = "v2_44", feature = "dox"))]
+    fn read_all_async<'a, P: Into<Option<&'a Cancellable>>, Q: Fn(Vec<u8>, Option<Error>) + Send + 'static>(&self, buffer: Vec<u8>, io_priority: i32, cancellable: P, callback: Q);
```
@antoyo might have an opinion on this one for the futures-glib stuff. I didn't look at that, but I assume you require the closures to get a `Result` passed, which is then converted into `Future::Item` and `Future::Error`? That might be an argument for making this a `Result<Vec<u8>, (Vec<u8>, Error)>` instead.

And generally, having a common "shape" for the async functions could be beneficial in the future for some generic higher-level thing around them.

EPashkin replied:
Minor question here what better `(Error, Vec<u8>)` or `(Vec<u8>, Error)` ?

sdroege replied:
Hm... or maybe even `Result<(Vec<u8>, Option<Error>), Error>`? The error case with read bytes is not *really* an error

EPashkin replied:
Then when we return `Err(err)` part?

sdroege replied:
When 0 bytes were read and there was an error

EPashkin replied:
IMHO we need return buffer in all error cases  to reuse be possible.
So `Result<Vec<u8>, (Vec<u8>, Error)>` or with reversed error tuple is only options.

sdroege replied:
`Result<(Vec<u8>, Option<Error>), (Vec<u8>, Error)>`?

EPashkin replied:
Maybe. So `read_all` need same change

sdroege commented:
Maybe also an async read test for the memory input stream could be useful. basically you could do it similar to the MainContext::invoke() test: create a main loop, "schedule" the read (i.e. call read_all_async()) and from the closures check the read data and quit the main loop. And after the call to read_all_async() you would run the main loop.

EPashkin commented:
Thanks, I was no idea how to do async test, and planned to look gtk/glib source to get some hints.

EPashkin commented:
@sdroege, @GuillaumeGomez I have question about `MemoryOutputStream`.
There one function that IMHO need for this is [`get_data`](https://developer.gnome.org/gio/stable/GMemoryOutputStream.html#g-memory-output-stream-get-data).
We can done just `Deref<Target=[u8]>` for stream,
but I prefer add `with_data(FnOnce(data: &[u8]))` instead, as data "may become invalid on the next write or truncate operation on the stream"
This can add problem for usage but add safety.
Maybe there other way?

GuillaumeGomez commented:
Can't we just have a specific type which will handle with corresponding lifetime?

sdroege commented:
The problem is that the memory output stream is reference counted. You could just have two references and then unsafety happens.

sdroege commented:
`get_data` needs to copy, everything else is unsafe. `steal_data/bytes` is fine and does not need to copy

EPashkin commented:
Problem with `steal_data` is "when no longer needed it must be freed using the free function set in ostream 's “destroy-function” property".
So I think better ignore it and  and use `steal_as_bytes` instead.

EPashkin commented:
Only one problem with `steal_as_bytes` that it one time operation, so it won't allow "peek" stream between writes.

sdroege commented:
You could add a `fn peek_data<F: FnOnce(&[u8])>(&self, func: F)` that calls `get_data()` and uses `slice::from_raw_parts()`

EPashkin commented:
Yes, is that I meant by my `with_data`, but now I don't sure that it not invalidated while closure called.

sdroege commented:
Ah indeed, you could pass a (second) reference of the stream into the closure and then cause unsafety.

sdroege commented:
Only `steal_as_bytes` it is then :)

EPashkin commented:
@sdroege Added tests for `read_all_async` with old `FnOnce(Vec<u8>, Option<Error>)` will change for `Result<(Vec<u8>, Option<Error>), (Vec<u8>, Error)>` later.
Take look please.

EPashkin commented:
There some results: 
1) async functions with `io_priority` need be `glib::Priority` instead `i32`.
2) Streams need at minimum `Send`, not sure that it right.

sdroege commented:
1) yes
2) why do you need them Send? But yes, and that's unsafe :( they are Send but not Sync, but they can't be Send either because reference counting

EPashkin commented:
1. Ok, I start replace it.
2. Thought use in tests but seems this unneeded

EPashkin commented on src/input_stream.rs:
```diff
@@ -0,0 +1,270 @@
+        }
+    }
+
+    fn read_async<'a, P: Into<Option<&'a Cancellable>>, Q: FnOnce(Result<Vec<u8>, (Vec<u8>, Error)>) + Send + 'static>(&self, mut buffer: Vec<u8>, io_priority: Priority, cancellable: P, callback: Q) {
```
@sdroege Can you check this function, tests with it failing with AV.
Seems I made some mistake, but can't find it.

EPashkin commented on src/input_stream.rs:
```diff
@@ -0,0 +1,270 @@
+        unsafe extern "C" fn read_async_trampoline<Q: FnOnce(Result<Vec<u8>, (Vec<u8>, Error)>) + Send + 'static>(_source_object: *mut gobject_ffi::GObject, res: *mut ffi::GAsyncResult, user_data: glib_ffi::gpointer)
+        {
+            callback_guard!();
+            let mut user_data: Box<Option<(Q, Vec<u8>)>> = Box::from_raw(user_data as *mut _);
```
Sorry, found that i forgot `Box` over `Q`

sdroege commented:
All good and working now and just needs another review, or was there an open question?

EPashkin commented:
@sdroege Can you look to `src/inputstream.rs`, please, when you have time? Maybe something in it can be done better.

sdroege commented:
Seems fine, thanks :) I'll go through all the other things another time tomorrow

sdroege commented:
Seems good overall, but I'm sure there are some hidden issues left. It's just so much code :) Let's fix those when they appear

We merge this after https://github.com/gtk-rs/gir/pull/509 so that the autogenerated code is also ok?

EPashkin commented:
This PR still long to go.
I still don't do anything with writers asyncs (`FnOnce`, buffers passtrought, tests).
`Converter` need `AsRef` and `AsMut`, etc.

Theoretically all async need manual fix to use `Priority`, but as it can't be source of error, I currently prefer skip this for functions with `FnOnce(Result<(), Error>)` callback.

sdroege commented:
You could detect priority in gir and rewrite it there for async functions

EPashkin commented:
As `Priority` need `.to_glib()` conversion IMHO this will difficult.

sdroege commented:
ToGlib is no problem, FromGlib would be

sdroege commented:
I was also thinking about the "read problem", where nothing is preventing the implementation to modify parts outside the actually read part. The same would also apply to the Rust Read trait for example. So maybe it's save to use any AsMut there and just return the number of read bytes, or rather it's as safe as the Read trait already is.

EPashkin commented:
For `MemoryOutputStream` IMHO better hide all except `new_resizable`, `get_data_size`, `steal_as_bytes` and `connect_property_data_size_notify`.
@GuillaumeGomez, @sdroege Any objections?

GuillaumeGomez commented:
None from me.

sdroege commented:
Makes sense, yes

EPashkin commented:
Cleared `MemoryOutputStream`
On starting processing `OutputStream` found that `write_async` also need buffer pass-through.
@sdroege Please, check it.

sdroege commented:
👍

sdroege commented on src/buffered_input_stream.rs:
```diff
@@ -0,0 +1,50 @@
+use Error;
+
+pub trait BufferedInputStreamExtManual {
+    fn fill_async<'a, P: Into<Option<&'a Cancellable>>, Q: Fn(Result<isize, Error>) + Send + Sync + 'static>(&self, count: isize, io_priority: i32, cancellable: P, callback: Q);
```
This is still using the old closures

sdroege commented on src/file_i_o_stream.rs:
```diff
@@ -0,0 +1,71 @@
+
+    fn query_info<'a, P: Into<Option<&'a Cancellable>>>(&self, attributes: &str, cancellable: P) -> Result<FileInfo, Error>;
+
+    fn query_info_async<'a, P: Into<Option<&'a Cancellable>>, Q: Fn(Result<FileInfo, Error>) + Send + Sync + 'static>(&self, attributes: &str, io_priority: i32, cancellable: P, callback: Q);
```
And this

sdroege commented on src/file_input_stream.rs:
```diff
@@ -0,0 +1,63 @@
+pub trait FileInputStreamExt {
+    fn query_info<'a, P: Into<Option<&'a Cancellable>>>(&self, attributes: &str, cancellable: P) -> Result<FileInfo, Error>;
+
+    fn query_info_async<'a, P: Into<Option<&'a Cancellable>>, Q: Fn(Result<FileInfo, Error>) + Send + Sync + 'static>(&self, attributes: &str, io_priority: i32, cancellable: P, callback: Q);
```
And this

sdroege commented on src/file_output_stream.rs:
```diff
@@ -0,0 +1,71 @@
+
+    fn query_info<'a, P: Into<Option<&'a Cancellable>>>(&self, attributes: &str, cancellable: P) -> Result<FileInfo, Error>;
+
+    fn query_info_async<'a, P: Into<Option<&'a Cancellable>>, Q: Fn(Result<FileInfo, Error>) + Send + Sync + 'static>(&self, attributes: &str, io_priority: i32, cancellable: P, callback: Q);
```
And this

sdroege commented:
@EPashkin Btw, did you see my comment here yesterday https://github.com/gtk-rs/gio/pull/69#issuecomment-351369144 ?

EPashkin commented:
Missed it. I will try to use `AsMut` in `InputStream` too.
So currently only fully processed memory streams, all others need second pass.

EPashkin commented:
@sdroege  We forgotten that `AsMut` impossible for reads until we use `.set_len`.
Change tactic and return buffer and count filled? Or leave as is.

EPashkin commented:
I meant, that we need use initialized buffer and pass `len` instead `capacity`?

sdroege commented:
That was my point, you can actually use `AsMut<[u8]>` here and return the number of read bytes. It's basically the same as the `Read` trait. If that is considered safe, we can consider it safe here too.

EPashkin commented:
Ok, I do this tomorrow, too late for today :(

sdroege commented:
Sure, no problem :)

EPashkin commented:
Now `src/input_stream.rs` and `src/output_stream.rs` is done.
Others streams still not fully.

sdroege commented:
:+1:

EPashkin commented:
Seems [Simplify test_util::run_async](https://github.com/EPashkin/gio/commit/5da2f2457571e3a3715eb350fc731350798eb4e1) was bad idea

sdroege commented on src/file.rs:
```diff
@@ -13,246 +13,225 @@ use FileQueryInfoFlags;
 
 pub trait FileExtManual {
-    fn append_to_async<'a, P: Into<Option<&'a Cancellable>>, Q: Fn(Result<FileOutputStream, Error>) + Send + Sync + 'static>(&self, flags: FileCreateFlags, io_priority: i32, cancellable: P, callback: Q);
+    fn append_to_async<'a, P: Into<Option<&'a Cancellable>>, Q: FnOnce(Result<FileOutputStream, Error>) + Send + 'static>(&self, flags: FileCreateFlags, io_priority: Priority, cancellable: P, callback: Q);
```
Why are *all* these manual? Only because of the priority? I'd rather add support to gir for this instead, this is a huge amount of manual code otherwise :)

EPashkin replied:
I think that it part of https://github.com/gtk-rs/gir/issues/504

sdroege replied:
Not really, but it's a missing feature in gir indeed. It's a normal GIO-style async function not handled by gir yet

EPashkin commented:
It because generated have `Result<(),Error>` instead needed

sdroege commented:
> It because generated have Result<(),Error> instead needed

Oh why? `g_file_append_to_finish()` returns a `FileOutputStream`

EPashkin commented:
@GuillaumeGomez , @sdroege 
PR is ready for final review.

sdroege commented on Gir.toml:
```diff
@@ -43,7 +41,6 @@ generate = [
     "Gio.SettingsBindFlags",
     "Gio.SettingsSchema",
     "Gio.SimpleActionGroup",
-    "Gio.SimpleIOStream",
```
Why do you drop this one? It's not very important for bindings in any case

EPashkin replied:
It just moved as object to hide properties

sdroege commented on src/auto/filter_input_stream.rs:
```diff
@@ -31,8 +31,6 @@ pub trait FilterInputStreamExt {
 
     fn set_close_base_stream(&self, close_base: bool);
 
-    fn connect_property_base_stream_notify<F: Fn(&Self) + 'static>(&self, f: F) -> SignalHandlerId;
```
I.e. generate = ["getter", "setter", "notify", "emit"]

EPashkin replied:
Good idea

EPashkin replied:
Added issue https://github.com/gtk-rs/gir/issues/513

sdroege commented:
Seems good for me, I'm just worried by the huge amount of manual code and the inconsistencies around the `io_priority` stuff. Maybe let's get it in, and then make sure that gir gets fixed for handling most of these cases (i.e. all but the read/write functions)

EPashkin commented:
@GuillaumeGomez please, look at this PR too.

GuillaumeGomez commented:
Seems good to me as well, thanks!
